File: LeetCode/JavaScript/#226 Invert Binary Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */

// recursion
var invertTree = function(root) {
	if(root === null){
		return root;
	}
    
    let temp = root.left;
	root.left = invertTree(root.right);
	root.right = invertTree(temp);
    
	return root;
};

// iteration
var invertTree = function(root) {
	if(root==null){
		return root;
	}
	var s = [root];
	while(s.length!==0){
		var top = s[s.length-1];
		var tmp = top.left;
		top.left = top.right;
		top.right = tmp;
		s.pop();
		if(top.left!==null){
			s.push(top.left);
		}
		if(top.right!==null){
			s.push(top.right);
		}
	}
	return root;
};

File: LeetCode/Python/#226 Invert Binary Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


# recursion
class Solution(object):
    def invertTree(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """
        if root == None:
            return None
        (root.left,root.right) = (self.invertTree(root.right),self.invertTree(root.left))
        return root


# iteration
class Solution(object):
    def invertTree(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """
        if root == None:
            return None
        s = [root]
        while(s):
            tmp = s.pop()
            if tmp:
                tmp.left,tmp.right = tmp.right,tmp.left
                s.append(tmp.left)
                s.append(tmp.right)
        return root

